src/bech32.cpp

```diff
@@ -32,182 +33,72 @@ const int8_t CHARSET_REV[128] = {
-    925, 735, 431, 784, 54, 609, 604, 39, 812, 737, 729, 466, 14, 533, 958,
-    481, 770, 499, 855, 238, 182, 464, 569, 72, 947, 442, 642, 24, 87, 989,
-    688, 88, 47, 762, 623, 709, 455, 817, 526, 637, 258, 84, 845, 738, 768,
-    698, 423, 933, 664, 620, 607, 629, 212, 347, 249, 982, 935, 131, 89, 252,
-    927, 189, 788, 853, 237, 691, 646, 403, 1010, 734, 253, 874, 807, 903,
-    1020, 100, 802, 71, 799, 1003, 633, 355, 276, 300, 649, 64, 306, 161,
-    608, 496, 743, 180, 485, 819, 383, 1016, 226, 308, 393, 648, 107, 19, 37,
-    585, 2, 175, 645, 247, 527, 5, 419, 181, 317, 327, 519, 542, 289, 567,
-    430, 579, 950, 582, 994, 1021, 583, 234, 240, 976, 41, 160, 109, 677,
-    937, 210, 95, 959, 242, 753, 461, 114, 733, 368, 573, 458, 782, 605, 680,
-    544, 299, 73, 652, 905, 477, 690, 93, 824, 882, 277, 946, 361, 17, 945,
-    523, 472, 334, 930, 597, 603, 793, 404, 290, 942, 316, 731, 270, 960,
-    936, 133, 122, 821, 966, 679, 662, 907, 282, 968, 767, 653, 20, 697, 222,
-    164, 835, 30, 285, 886, 456, 436, 640, 286, 1015, 380, 840, 245, 724,
-    137, 593, 173, 130, 715, 85, 885, 551, 246, 449, 103, 366, 372, 714, 313,
-    865, 241, 699, 674, 374, 68, 421, 562, 292, 59, 809, 342, 651, 459, 227,
-    46, 711, 764, 868, 53, 413, 278, 800, 255, 993, 318, 854, 319, 695, 315,
-    469, 166, 489, 969, 730, 1001, 757, 873, 686, 197, 303, 919, 155, 673,
-    940, 712, 25, 999, 63, 863, 972, 967, 785, 152, 296, 512, 402, 377, 45,
-    899, 829, 354, 77, 69, 856, 417, 811, 953, 124, 418, 75, 794, 162, 414,
-    1018, 568, 254, 265, 772, 588, 16, 896, 157, 889, 298, 621, 110, 844,
-    1000, 108, 545, 601, 78, 862, 447, 185, 195, 818, 450, 387, 49, 805, 102,
-    986, 1005, 827, 329, 28, 932, 410, 287, 435, 451, 962, 517, 48, 174, 43,
-    893, 884, 261, 251, 516, 395, 910, 611, 29, 501, 223, 476, 364, 144, 871,
-    998, 687, 928, 115, 453, 513, 176, 94, 168, 667, 955, 353, 434, 382, 400,
-    139, 365, 996, 343, 948, 890, 1012, 663, 610, 718, 538, 1008, 639, 470,
-    848, 543, 1011, 859, 671, 756, 83, 427, 159, 746, 669, 589, 971, 524,
-    356, 995, 904, 256, 201, 988, 62, 397, 81, 720, 917, 209, 549, 943, 486,
-    76, 148, 207, 509, 644, 386, 700, 534, 177, 550, 961, 926, 546, 428, 284,
-    127, 294, 8, 269, 359, 506, 445, 997, 806, 591, 725, 178, 262, 846, 373,
-    831, 504, 305, 843, 553, 378, 1017, 783, 474, 683, 581, 200, 498, 694,
-    191, 217, 847, 941, 424, 235, 38, 74, 616, 786, 147, 4, 273, 214, 142,
-    575, 992, 463, 983, 243, 360, 970, 350, 267, 615, 766, 494, 31, 1009,
-    452, 710, 552, 128, 612, 600, 275, 322, 193
-};
-static_assert(std::size(GF1024_LOG) == 1024, "GF1024_EXP length should be 1024");
+/** We work with the finite field GF(1024) defined as a degree 2 extension of the base field GF(32)
+ * The defining polynomial of the extension is x^2 + 9x + 23
+ * Let (e) be a primitive element of GF(1024), that is, a generator of the field.
```

Comment:
This may be a bit of a philosophical point, touching on difference between specific field and field definition up to isomorphism.

Still, I don't think it's sufficient (or at least confusing) to say it is *a* primitive element of GF(1024). It is specifically one of the roots of `x^2 + 9x + 23`.

Reply:
How about this?

```
Define F=GF(32) as the set of polynomials over GF(2) modulo x^5 + x^3 + 1. We will represent elements of F as their integer encoding (e.g. 11 to mean x^3 + x + 1).
Define E=GF(1024) as an extension of F, namely the set of polynomials over F modulo x^2+ 9x + 23. We will represent the element (ax + b) of F as the integer 32*a+b. Let e be the primitive element x of F (represented as 32).
Every non-zero element of the field can then ...
```

Reply:
I think this is the definition of primitive element. At least, that agrees with what Wikipedia says:

> In field theory, a primitive element of a finite field GF(q) is a generator of the multiplicative group of the field. In other words, α ∈ GF(q) is called a primitive element if it is a primitive (q − 1)th root of unity in GF(q); this means that each non-zero element of GF(q) can be written as α^i for some integer i.

But I agree this is more than just any primitive element, it is specifically because it is a root of the defining polynomial that we decide to use it. I will amend the comments.